### PR TITLE
Specialize function wrapping to preserve arg len

### LINKED ISSUE #53: Requiring async listener results in incorrect function lengths.
Run with async-listener @0.5.7 and node v5.7.1.

```
> var fs = require('fs');
undefined
> fs.readFile.length
3
> require('async-listener')
{}
> fs.readFile.length
0
```

### PR DESCRIPTION
**Pull request by matthewloring**
Fixes #53

### CODE REVIEW

**Comment by matthewloring** [comment]
cc @Qard @hayes @faithseed @Jeff-Lewis

File: index.js [hunk 1]
```diff
@@ -460,3 +439,94 @@ function wrapPromise() {
...
+      };
+    case 2:
+      return function (a, cb) {
+        if (arguments.length !== 2) return fallback.apply(this, arguments);
+        if (typeof cb === "function") cb = wrapCallback(cb);
+        return fn.call(this, a, cb);
+      };
+    case 3:
+      return function (a, b, cb) {
+        if (arguments.length !== 2) return fallback.apply(this, arguments);
```
Comment (Qard) [inline index.js]: Some of those length checks need adjustment. :wink:

Comment (matthewloring) [inline index.js]: Oh dang, good catch!

**Comment by Qard** [comment]
LGTM, but I'd like to see another LGTM from someone else on this before merging.

**Comment by hayes** [comment]
I'll take a look tomorrow

**Comment by hayes** [comment]
I'll take a look tomorrow

File: index.js [hunk 2]
```diff
@@ -460,3 +439,94 @@ function wrapPromise() {
...
+      arguments[index] = wrapCallback(arguments[index]);
+    }
+    return fn.apply(this, arguments);
+  };
+  // Preserve function length for small arg count functions.
+  switch (fn.length) {
+    case 1:
+      return function (cb) {
+        if (arguments.length !== 1) return fallback.apply(this, arguments);
+        if (typeof cb === "function") cb = wrapCallback(cb);
```
Comment (hayes) [inline index.js]: I recall there being some additional overhead to redefining/assigning to argument variables, but I can't remember where I got that from.

**Comment by hayes** [comment]
Looks good to me

File: index.js (hunk 2 above)
Comment (Jeff-Lewis) [inline index.js]: Do you something like what this package refers to? https://www.npmjs.com/package/manage-arguments

Comment (matthewloring) [inline index.js]: It appears that reassigning argument variables is only a performance concern in non-strict mode. See: https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#3-managing-arguments

Comment (hayes) [inline index.js]: Cool, thanks for finding that.

**Comment by matthewloring** [comment]
Is there any further work you'd like on this before landing?

**Comment by Qard** [comment]
Ah, sorry, forgot to get back to this. Could you add a simple test just to verify the functions have the appropriate lengths?

**Comment by matthewloring** [comment]
@Qard I've added a test

**Comment by Qard** [comment]
merged and released as 0.6.0 due to the minor behavioural difference.

**Comment by matthewloring** [comment]
Thanks!

**Pull request merged**